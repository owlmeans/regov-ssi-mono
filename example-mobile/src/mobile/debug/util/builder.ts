/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Config } from "@owlmeans/regov-lib-react"
import { webCryptoHelper, buildWalletWrapper } from "@owlmeans/regov-ssi-core"


export const buildDevWallet = async (config: Config, alias = DEFAULT_DEVELOPMENT_VOICE_ALIAS) =>
  await buildWalletWrapper(
    webCryptoHelper, '11111111', { alias, name: 'Development wallet' }, { 
      prefix: config.DID_PREFIX,
      defaultSchema: config.baseSchemaUrl,
      didSchemaPath: config.DID_SCHEMA_PATH,
    }
  )

export const DEFAULT_DEVELOPMENT_VOICE_ALIAS = 'development'