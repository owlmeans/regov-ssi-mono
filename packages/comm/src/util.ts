/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { generateKeyPairFromSeed } from '@stablelib/x25519'
import {
  CryptoKey, DIDDocument, DIDHelper, DIDPURPOSE_AGREEMENT, KEYCHAIN_ERROR_NO_KEY, KeyPairToCryptoKeyOptions
} from "@owlmeans/regov-ssi-core"
import { CommKey, COMM_DID_AGREEMENT_KEY_DEFAULT, connectionFieldList, DIDCommConnectMeta, ERROR_COMM_DID_NOKEY, ERROR_COMM_NO_RECIPIENT } from "./types"
import { JWE } from 'did-jwt'


export const cryptoKeyToCommKey = async (
  key: CryptoKey, options?: KeyPairToCryptoKeyOptions
): Promise<CommKey> => {
  if (!key.pk) {
    throw Error(KEYCHAIN_ERROR_NO_KEY)
  }
  const pk = Buffer.from(key.pk, 'utf8').subarray(0, 32)
  const pubKey = generateKeyPairFromSeed(pk).publicKey

  return {
    id: options?.id || key.id || '_noid_',
    pk, pubKey
  }
}

export const filterConnectionFields = (connection: DIDCommConnectMeta): DIDCommConnectMeta => {
  return Object.fromEntries(
    connectionFieldList.map(field => [field, connection[field]])
  ) as unknown as DIDCommConnectMeta
}

export const invertConnection = (connection: DIDCommConnectMeta, channel?: string): DIDCommConnectMeta => {
  const newConnection = filterConnectionFields(JSON.parse(JSON.stringify(connection)))

  if (!newConnection.recipient) {
    throw new Error(ERROR_COMM_NO_RECIPIENT)
  }

  return {
    ...newConnection,
    recipientId: newConnection.sender.id,
    recipient: newConnection.sender,
    sender: newConnection.recipient,
    channel: channel || newConnection.channel
  }
}

export const parseJWE = (data: string) => {
  try {
    return JSON.parse(data) as JWE
  } catch (e) {
  }

  return undefined
}

export const didDocToCommKeyBuilder = (helper: DIDHelper) =>
  async (did: DIDDocument, keyId: string = COMM_DID_AGREEMENT_KEY_DEFAULT): Promise<Partial<CommKey>> => {
    const agreement = helper.expandVerificationMethod(did, DIDPURPOSE_AGREEMENT, keyId)
    if (!agreement.publicKeyBase58) {
      throw new Error(ERROR_COMM_DID_NOKEY)
    }

    return {
      id: did.id,
      pubKey: helper.getCrypto().base58().decode(agreement.publicKeyBase58)
    }
  }