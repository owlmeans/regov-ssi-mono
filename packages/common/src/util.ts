import { MaybeArray, Idish } from "./types"

export const normalizeValue = <Type>(value: MaybeArray<Type>): Type[] => {
  return Array.isArray(value)
    ? value
    : value ? [value] : []
}

export const simplifyValue = <Type>(value: MaybeArray<Type>) => {
  return Array.isArray(value) && value.length < 2
    ? value.length === 0 ? undefined : value[0]
    : value
}

export const singleValue = <Type>(value: MaybeArray<Type>): Type | undefined => {
  return Array.isArray(value)
    ? value.length > 0 ? value[0] : undefined
    : value
}

export const mapValue = async <Type, Res>(
  value: MaybeArray<Type>,
  callback: (value: Type) => Promise<Res>
): Promise<Res[]> => {
  return Promise.all(normalizeValue(value).map(callback))
}

export const addToValue = <Type>(value: MaybeArray<Type> | undefined, add: MaybeArray<Type>): MaybeArray<Type> => {
  if (value) {
    return [...normalizeValue(value), ...normalizeValue(add)]
  }

  return add
}

export const extractId = (id: Idish) => typeof id === 'string' ? id : id.id

export const getDeepValue = <Type, Subject extends {}>(subject: Subject, key: string, def?: Type): Type => {
  const parts = key.split('.')
  return parts.reduce((value, key) => {
    if (subject.hasOwnProperty(key)) {
      const tmp = (subject as any)[key]
      subject = tmp

      return tmp
    }

    return def
  }, def) as Type
}