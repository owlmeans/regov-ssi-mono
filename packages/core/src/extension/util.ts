/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, normalizeValue } from "../common"
import { BasicCredentialType, Credential } from "../vc"
import {
  CredentialService, CredentialServiceBuilder, defaultBuildMethod, defaultClaimMethod,
  defaultOfferMethod, defaultRequestMethod, defaultRespondMethod, defaultSignMethod, defaultValidateMethod,
  Extension,
  ValidationResult
} from "./ext"
import { CredentialDescription } from "./schema"


export const findAppropriateCredentialType = (
  ext: Extension, types: BasicCredentialType, defaultType: string
): string => {
  types = normalizeValue(types)

  return (Object.entries(ext.factories).map(([type]) => type)
    .find(type => types.includes(type)) || defaultType)
}

const methodMap = {
  'produceBuildMethod': 'build',
  'produceSignMethod': 'sign',
  'produceValidateMethod': 'validate',
  'produceClaimMethod': 'claim',
  'produceOfferMethod': 'offer',
  'produceRequestMethod': 'request',
  'produceRespondMethod': 'respond'
}

export const addFactoriesToExt = (ext: Extension, type: string, factories: CredentialServiceBuilder) => {
  const description = (ext.schema.credentials as { [key: string]: CredentialDescription })[type]
  ext.factories[type] = {
    ...{
      build: defaultBuildMethod(description),
      sign: defaultSignMethod(description),
      validate: defaultValidateMethod(description),
      claim: defaultClaimMethod(description),
      offer: defaultOfferMethod(description),
      request: defaultRequestMethod(description),
      respond: defaultRespondMethod(description)
    },
    ...Object.entries(factories).reduce((_facts, [method, builder]) => {
      if (ext.schema.credentials) {
        return {
          ..._facts,
          [(methodMap as any)[method]]: builder(description)
        }
      }
      return _facts
    }, {} as CredentialService)
  }
}

interface CredentialChainDebug {
  id: string
  type: string[]
  children: CredentialChainDebug[]
}

export const debugCredentialChain = (credential: Credential): CredentialChainDebug => {
  const result: CredentialChainDebug = { id: credential.id, type: credential.type, children: [] }

  result.children = normalizeValue(credential.evidence).map(
    evidence => debugCredentialChain(evidence as Credential)
  )

  return result
}

export const debugValidationResult = (result: ValidationResult): CredentialChainDebug => {
  const _result: CredentialChainDebug = {
    id: result.instance?.id || '',
    type: result.instance?.type || [''],
    children: []
  }

  _result.children = normalizeValue(result.evidence).map(
    evidence => debugValidationResult(evidence.result)
  )

  return _result
}