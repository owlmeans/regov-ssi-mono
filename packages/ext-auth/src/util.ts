/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Credential, Presentation } from "@owlmeans/regov-ssi-core";
import { 
  AuthSubject, REGOV_AUTH_REQUEST_TYPE, REGOV_AUTH_RESPONSE_TYPE, REGOV_CREDENTIAL_TYPE_AUTH 
} from "./types";


export const getAuthFromPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(REGOV_AUTH_RESPONSE_TYPE)
   && !presentation.type.includes(REGOV_AUTH_REQUEST_TYPE)) {
    return undefined
  }

  return presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_CREDENTIAL_TYPE_AUTH)
  )
}

export const getAuthSubject = (cred: Credential): AuthSubject => {
  return cred.credentialSubject as unknown as AuthSubject
}

export const pinValidation = {
  required: true,
  minLength: 4,
  pattern: /^\d+$/
}