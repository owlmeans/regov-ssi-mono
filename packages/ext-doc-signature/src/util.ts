import { Presentation } from "@owlmeans/regov-ssi-core"
import { REGOV_SIGNATURE_REQUEST_TYPE } from "./types"


export const getSignatureRequestFromPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(REGOV_SIGNATURE_REQUEST_TYPE)) {
    return undefined
  }

  return presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_SIGNATURE_REQUEST_TYPE)
  )
}