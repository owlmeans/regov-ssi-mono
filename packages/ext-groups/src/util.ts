/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, normalizeValue } from "@owlmeans/regov-ssi-core"
import { Presentation, Credential } from "@owlmeans/regov-ssi-core"
import { DIDDocument } from "@owlmeans/regov-ssi-core"
import {
  REGOV_CREDENTIAL_TYPE_GROUP, REGOV_CREDENTIAL_TYPE_MEMBERSHIP, REGOV_CLAIM_TYPE,
  BASIC_IDENTITY_TYPE, REGOV_OFFER_TYPE
} from "./types"


export const getGroupFromMembershipClaimPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(REGOV_CLAIM_TYPE)) {
    return undefined
  }

  const membershipClaim = presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_CREDENTIAL_TYPE_MEMBERSHIP)
  )

  return normalizeValue(membershipClaim?.evidence).find(
    evidence => evidence?.type.includes(REGOV_CREDENTIAL_TYPE_GROUP)
  ) as Credential
}

export const getGroupOwnerIdentity = (crednetial: Credential) => {
  if (!crednetial.type.includes(REGOV_CREDENTIAL_TYPE_GROUP)) {
    return undefined
  }

  return normalizeValue<Credential>(crednetial.evidence as MaybeArray<Credential>).find(
    evidence => evidence?.type.includes(BASIC_IDENTITY_TYPE)
      && evidence.id === (crednetial.issuer as unknown as DIDDocument).id
  )
}

export const getMembershipClaimHolder = (presentation: Presentation) => {
  const membership = presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_CREDENTIAL_TYPE_MEMBERSHIP)
  )

  return membership?.issuer as unknown as DIDDocument
}

export const getMembershipClaim = (presentation: Presentation) => {
  return presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_CREDENTIAL_TYPE_MEMBERSHIP)
  )
}

export const getGroupFromMembershipOfferPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(REGOV_OFFER_TYPE)) {
    return undefined
  }

  const membershipClaim = presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_CREDENTIAL_TYPE_MEMBERSHIP)
  )

  return normalizeValue(membershipClaim?.evidence).find(
    evidence => evidence?.type.includes(REGOV_CREDENTIAL_TYPE_GROUP)
  ) as Credential
}

export const getMembershipOffer = (presentation: Presentation) => {
  return presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_CREDENTIAL_TYPE_MEMBERSHIP)
  )
}