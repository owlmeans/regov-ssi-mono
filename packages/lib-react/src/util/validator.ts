/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */


export const validateJsonOrEmpty = (v: string) => {
  if (v === '') {
    return true
  }

  return validateJson(v)
}

export const validateJson = (v: string) => {
  try {
    const tmp = JSON.parse(v)
    if (typeof tmp === 'object') {
      return true
    }
  } catch (e) {
    return false
  }

  return true
}

export const loginAliasValidation = {
  required: true,
  minLength: 3,
  maxLength: 24,
  pattern: /^[\w\d_-]+$/
}

export const passwordValidation = {
  required: true,
  minLength: 8,
  maxLength: 64
}

export const didValidation = {
  required: true,
  minLength: 10,
  maxLength: 128,
  pattern: /^did\:\w+\:.+/
}

export const humanReadableVersion = {
  maxLength: 128,
  pattern: /^[\w\d\s\._-]*$/
}

export const generalNameVlidation = (required: boolean = true) => ({
  required,
  maxLength: 96,
  validate: {
    pattern: (v: string) => !v.match(/[\<\>\[\]\{\}\\\']/)
  }
})
export const generalIdVlidation = (required: boolean = true) => ({
  required,
  maxLength: 1024
})

export const urlVlidation = (required: boolean = false) => ({
  validate: {
    url: (v: string) => {
      if (!required && v === '') {
        return true
      }
      
      try {
        new URL(v)
      } catch (e) {
        return false
      }

      return true
    }
  }
})

