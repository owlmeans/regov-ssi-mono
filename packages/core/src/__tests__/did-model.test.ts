/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

require('dotenv').config()

import {
  buildDidHelper, nodeCryptoHelper, normalizeValue, DIDDocument, DIDPURPOSE_ASSERTION, 
  DIDPURPOSE_VERIFICATION, buildDidRegistryWarpper, VERIFICATION_KEY_CONTROLLER
} from "../index"

import util from 'util'
util.inspect.defaultOptions.depth = 8


const ctx: {
  didDoc?: DIDDocument
  holderDoc?: DIDDocument
} = {}

describe('DID Helper', () => {
  const didHelper = buildDidHelper(nodeCryptoHelper)
  buildDidRegistryWarpper(didHelper)

  it('creates DID Id', async () => {
    const key = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))
    const id = didHelper.makeDIDId(key)
    const described = id.split(':')
    expect(described.length).toBe(3)
    expect(described).toContain('did')
    expect(described).toContain(process.env.DID_PREFIX)
    expect(typeof described[2]).toBe('string')
  })

  it('creates a DID Document', async () => {
    const key = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))

    const didDocUnsinged = await didHelper.createDID(key, {
      purpose: [DIDPURPOSE_VERIFICATION, DIDPURPOSE_ASSERTION],
      data: 'Hello world!',
      hash: true
    })

    const didDoc = await didHelper.signDID(key, didDocUnsinged)

    ctx['didDoc'] = didDoc
    expect(didDoc).toMatchSnapshot({
      id: expect.any(String),
      verificationMethod: {
        id: expect.any(String),
        nonce: expect.any(String),
        controller: expect.any(String),
        publicKeyBase58: expect.any(String),
      },
      assertionMethod: expect.any(String),
      proof: {
        created: expect.any(String),
        jws: expect.any(String),
        verificationMethod: expect.any(String),
      }
    })
  })

  it('verifies DID Document from itself', async () => {
    if (!ctx.didDoc) {
      throw new Error('No DID doc from pervious test')
    }
    const result = await didHelper.verifyDID(ctx.didDoc)

    expect(result).toBe(true)
  })

  it('produces holder / controller verifiable did', async () => {
    const aliceKey = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))
    aliceKey.nextKeyDigest = 'nkdigest-simulation'
    const bobKey = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))
    bobKey.nextKeyDigest = 'nkdigest-simulation'

    const didDocUnsinged = await didHelper.createDID(aliceKey, {
      purpose: [DIDPURPOSE_VERIFICATION, DIDPURPOSE_ASSERTION],
      data: 'Hello world!',
      hash: true
    })

    const didDoc = await didHelper.signDID(
      bobKey,
      didDocUnsinged,
      VERIFICATION_KEY_CONTROLLER,
      [DIDPURPOSE_ASSERTION]
    )

    ctx.holderDoc = didDoc
  })

  it('verifies holder veriable did', async () => {
    if (!ctx.holderDoc) {
      throw new Error('No DID doc from pervious test')
    }
    const result = await didHelper.verifyDID(ctx.holderDoc)

    expect(result).toBe(true)
  })

  it('fails on tempered data', async () => {
    if (!ctx.holderDoc) {
      throw new Error('No DID doc from pervious test')
    }

    const brokenDoc = <DIDDocument>JSON.parse(JSON.stringify(ctx.holderDoc))
    if (brokenDoc.verificationMethod && normalizeValue(brokenDoc.verificationMethod).length > 0
      && typeof brokenDoc.proof === 'object') {
      brokenDoc.proof.created = new Date().toUTCString()
    }
    const result = await didHelper.verifyDID(brokenDoc)

    expect(result).toBe(false)
  })
})