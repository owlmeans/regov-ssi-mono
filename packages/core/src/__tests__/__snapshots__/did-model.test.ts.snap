// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DID Helper creates a DID Document 1`] = `
Object {
  "@context": Array [
    "https://w3id.org/did/v1",
    "https://owlmeans.com/schemas/did-schema.json#",
  ],
  "assertionMethod": Any<String>,
  "id": Any<String>,
  "proof": Object {
    "created": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "verificationMethod",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "verificationMethod": Object {
    "controller": Any<String>,
    "id": Any<String>,
    "nonce": Any<String>,
    "publicKeyBase58": Any<String>,
    "type": "EcdsaSecp256k1VerificationKey2019",
  },
}
`;
